src/typechecking.rs

```diff
@@ -36,57 +36,47 @@ impl Default for TypeInfo {
             global_types: Default::default(),
         };
 
-        res.add_sort(UnitSort::new(UNIT_SYM.into()), DUMMY_SPAN.clone());
-        res.add_sort(StringSort::new("String".into()), DUMMY_SPAN.clone());
-        res.add_sort(BoolSort::new("bool".into()), DUMMY_SPAN.clone());
-        res.add_sort(I64Sort::new("i64".into()), DUMMY_SPAN.clone());
-        res.add_sort(F64Sort::new("f64".into()), DUMMY_SPAN.clone());
-        res.add_sort(RationalSort::new("Rational".into()), DUMMY_SPAN.clone());
+        res.add_sort(UnitSort, DUMMY_SPAN.clone()).unwrap();
+        res.add_sort(StringSort, DUMMY_SPAN.clone()).unwrap();
+        res.add_sort(BoolSort, DUMMY_SPAN.clone()).unwrap();
+        res.add_sort(I64Sort, DUMMY_SPAN.clone()).unwrap();
+        res.add_sort(F64Sort, DUMMY_SPAN.clone()).unwrap();
+        res.add_sort(RationalSort, DUMMY_SPAN.clone()).unwrap();
 
-        res.presort_names.extend(MapSort::presort_names());
-        res.presort_names.extend(SetSort::presort_names());
-        res.presort_names.extend(VecSort::presort_names());
-        res.presort_names.extend(FunctionSort::presort_names());
+        res.add_presort::<MapSort>(DUMMY_SPAN.clone()).unwrap();
+        res.add_presort::<SetSort>(DUMMY_SPAN.clone()).unwrap();
+        res.add_presort::<VecSort>(DUMMY_SPAN.clone()).unwrap();
+        res.add_presort::<FunctionSort>(DUMMY_SPAN.clone()).unwrap();
 
-        res.presorts.insert("Map".into(), MapSort::make_sort);
-        res.presorts.insert("Set".into(), SetSort::make_sort);
-        res.presorts.insert("Vec".into(), VecSort::make_sort);
-        res.presorts
-            .insert("UnstableFn".into(), FunctionSort::make_sort);
-
-        res.add_primitive(ValueEq {
-            unit: res.get_sort_nofail(),
-        });
+        res.add_primitive(ValueEq);
 
         res
     }
 }
 
-pub const UNIT_SYM: &str = "Unit";
-
 impl TypeInfo {
-    pub(crate) fn infer_literal(&self, lit: &Literal) -> ArcSort {
-        match lit {
-            Literal::Int(_) => self.sorts.get(&Symbol::from("i64")),
-            Literal::F64(_) => self.sorts.get(&Symbol::from("f64")),
-            Literal::String(_) => self.sorts.get(&Symbol::from("String")),
-            Literal::Bool(_) => self.sorts.get(&Symbol::from("bool")),
-            Literal::Unit => self.sorts.get(&Symbol::from("Unit")),
-        }
-        .unwrap()
-        .clone()
+    pub fn add_sort<S: Sort + 'static>(&mut self, sort: S, span: Span) -> Result<(), TypeError> {
+        self.add_arcsort(Arc::new(sort), span)
     }
 
-    pub fn add_sort<S: Sort + 'static>(&mut self, sort: S, span: Span) {
-        self.add_arcsort(Arc::new(sort), span).unwrap()
+    pub fn add_presort<S: Presort>(&mut self, span: Span) -> Result<(), TypeError> {
```

Comment:
I’m less sure about this interface. This means presort (unlike sort) has to be static, and we can’t define the behavior of a presort dynamically. There might be cases where we may want not only sort but presort to be dynamically defined in the future. Examples include presorts `(Database a)` or `(Network a)` that read values of type `a` from an external source, where the presort need to own a database/network object

Reply:
I don't believe that this trait prevents any of that stuff though? All I was trying to do was collect the methods that presorts already depend on in one place.

Reply:
Conclusion: add a comment to Presort explaining that it's currently intended to be used on the same struct as Sort but that can change if e.g. we want partial application in the future.